### ISSUE TITLE
WebDav xmldb and xmlrpc -> Moving files locked by another user is prevented directly but not via collections

### ISSUE THREAD

**Issue by tremms18** [eXist-db/exist#954]
I found that I am prevented from moving a file that is locked by another user if I try to move it by itself.  If I go to the same file's parent collection and move the collection, I am able to move the file via the collection.

**Comment by adamretter** [comment]
It looks like this effects the `WebDAV`, `XML-RPC` and `XML:DB` API's.

It seems that `UserLocks` are only checked in `NativeBroker#moveResource` and not `NativeBroker#moveCollection`.

`moveResource` looks has these lines:

```java
final Account docUser = doc.getUserLock(transaction);
if(docUser != null) {
    if(!(getCurrentSubject().getName()).equals(docUser.getName())) {
        throw new PermissionDeniedException("Cannot move '" + doc.getFileURI() + " because is locked by getUser() '" + docUser.getName() + "'");
    }
}
```

It seems to be that we should be consistent in the least, i.e. either:
1. Ignore the User Lock completely (which would seem silly)
2. Respect the User Lock in both `moveResource` and `moveCollection`

**Comment by adamretter** [comment]
@dizzzz As this effects the WebDAV API, what do you think?

**Comment by dizzzz** [comment]
I Will have a look later today....

**Comment by dizzzz** [comment]
Pooh, isn't a (kind of) fundamental issue? I realise that for this operation recursively must be checked if somewhere a lock is active (webdav is not special in this I guess, except that propound must report it too).

I wonder why we did not think of this before.

**Comment by adamretter** [comment]
@dizzzz Yeah to solve we have to either introduce a lock table or recursively search the sub-collection hierarchy for locked documents at Collection move time.

**Comment by dizzzz** [comment]
A locktable... Would be very helpful! I have been thinking to make one , when I implemented the WebDAV stuff. The way how locktokens are stored... Is sub optimal. 

A locktable would be awesome and fast!

**Comment by shabanovd** [comment]
@adamretter @dizzzz @wolfgangmm there is metadata storage, can it be useful for that? also, move current resource metadata will improve performance a lot.

**Comment by adamretter** [comment]
@shabanovd I am not really sure what you are proposing?

**Comment by dizzzz** [comment]
probably https://github.com/eXist-db/exist/tree/develop/extensions/metadata ...
https://github.com/eXist-db/exist/blob/develop/extensions/metadata/sleepycat/src/main/java/org/exist/storage/md/MetaDataImpl.java

**Comment by shabanovd** [comment]
@adamretter, yeap, @dizzzz is right. Metadata storage (extension) design to store key-value pair per resource (collections, xml & binary documents). It have simple methods to set & get that pairs. There is also backup integration.

**Comment by adamretter** [comment]
@shabanovd @dizzzz Sure I understood Dmitriy mean the metadata extension, but I don't see how that is relevant here?

**Comment by dizzzz** [comment]
My impression that he suggest not to built a separate lock-table, but to use this extension in stead; also it could be used to offload the current dbx files with (other) document meta-data.

**Comment by adamretter** [comment]
Hmm a lock table would be core to the database so I don't think an extension is a good idea. Also the lock-table does not need persistent storage.

**Comment by shabanovd** [comment]
@adamretter why metadata storage can't become core feature?

**Comment by adamretter** [comment]
@shabanovd It uses an external data-store, IMHO this doesn't make it easy in future to integrate transactions etc across both stores. Also, I don't see that the lock table needs to be persistent and so this seems like a lot of overhead for something simple.
The lock table will need to be fast though, so something like an in-memory skip list is probably enough.

**Comment by dizzzz** [comment]
i tend to agree with @adamretter here;

**Comment by shabanovd** [comment]
@adamretter "I don't see that the lock table needs to be persistent" ... hmmm, are you 100% sure on that?

**Comment by adamretter** [comment]
@shabanovd Nope ;-) So... why do you think we need to be able to persist it?

**Comment by shabanovd** [comment]
@adamretter because production server upgrade must not be visible for end user. All locks must persistent.

Also, eXist's metadata operations very heavy and must be optimized, making metadata storage core feature is good step for that too, IMHO

**Comment by dizzzz** [comment]
Ah I missed something crucial ; until now I have always been convinced that locks should be persistent (It adds additional complexity for sure).

But..... I need to think about the rationale about this; would it hurt if a lock is gone after server reset? not sure , @wolfgangmm ? @ljo ?

**Comment by dizzzz** [comment]
My initial idea is and was to add an new btree thingy to exist for metadata; all logic is their already no?

**Comment by shabanovd** [comment]
@dizzzz @adamretter @wolfgangmm @ljo Metadata storage is not just "add another btree", but question of infrastructure: backup/restore, handle copy, move, remove operations. Right now, metadata storage implemented with sleepycat storage, and it modular design allow to implement storage based on eXist's btree. Rest in place already.

**Comment by caseydawsonjordan** [comment]
Just wanted to chime in and add that the existing behavior of locks
persisting across restarts is pretty important to maintain. Users may lock
(checkout) files, and expect those locks to be maintained through restarts,
upgrades etc. For systems doing document management this is pretty crucial
behavior. Thanks!
